File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/InterpolationSyntax.cs
using System;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class InterpolationSyntax : InterpolatedStringContentSyntax
	{
		internal readonly PunctuationSyntax _openBraceToken;

		internal readonly ExpressionSyntax _expression;

		internal readonly InterpolationAlignmentClauseSyntax _alignmentClause;

		internal readonly InterpolationFormatClauseSyntax _formatClause;

		internal readonly PunctuationSyntax _closeBraceToken;

		internal static Func<ObjectReader, object> CreateInstance;

		internal PunctuationSyntax OpenBraceToken => _openBraceToken;

		internal ExpressionSyntax Expression => _expression;

		internal InterpolationAlignmentClauseSyntax AlignmentClause => _alignmentClause;

		internal InterpolationFormatClauseSyntax FormatClause => _formatClause;

		internal PunctuationSyntax CloseBraceToken => _closeBraceToken;

		internal InterpolationSyntax(SyntaxKind kind, PunctuationSyntax openBraceToken, ExpressionSyntax expression, InterpolationAlignmentClauseSyntax alignmentClause, InterpolationFormatClauseSyntax formatClause, PunctuationSyntax closeBraceToken)
			: base(kind)
		{
			base._slotCount = 5;
			AdjustFlagsAndWidth(openBraceToken);
			_openBraceToken = openBraceToken;
			AdjustFlagsAndWidth(expression);
			_expression = expression;
			if (alignmentClause != null)
			{
				AdjustFlagsAndWidth(alignmentClause);
				_alignmentClause = alignmentClause;
			}
			if (formatClause != null)
			{
				AdjustFlagsAndWidth(formatClause);
				_formatClause = formatClause;
			}
			AdjustFlagsAndWidth(closeBraceToken);
			_closeBraceToken = closeBraceToken;
		}

		internal InterpolationSyntax(SyntaxKind kind, PunctuationSyntax openBraceToken, ExpressionSyntax expression, InterpolationAlignmentClauseSyntax alignmentClause, InterpolationFormatClauseSyntax formatClause, PunctuationSyntax closeBraceToken, ISyntaxFactoryContext context)
			: base(kind)
		{
			base._slotCount = 5;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(openBraceToken);
			_openBraceToken = openBraceToken;
			AdjustFlagsAndWidth(expression);
			_expression = expression;
			if (alignmentClause != null)
			{
				AdjustFlagsAndWidth(alignmentClause);
				_alignmentClause = alignmentClause;
			}
			if (formatClause != null)
			{
				AdjustFlagsAndWidth(formatClause);
				_formatClause = formatClause;
			}
			AdjustFlagsAndWidth(closeBraceToken);
			_closeBraceToken = closeBraceToken;
		}

		internal InterpolationSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, PunctuationSyntax openBraceToken, ExpressionSyntax expression, InterpolationAlignmentClauseSyntax alignmentClause, InterpolationFormatClauseSyntax formatClause, PunctuationSyntax closeBraceToken)
			: base(kind, errors, annotations)
		{
			base._slotCount = 5;
			AdjustFlagsAndWidth(openBraceToken);
			_openBraceToken = openBraceToken;
			AdjustFlagsAndWidth(expression);
			_expression = expression;
			if (alignmentClause != null)
			{
				AdjustFlagsAndWidth(alignmentClause);
				_alignmentClause = alignmentClause;
			}
			if (formatClause != null)
			{
				AdjustFlagsAndWidth(formatClause);
				_formatClause = formatClause;
			}
			AdjustFlagsAndWidth(closeBraceToken);
			_closeBraceToken = closeBraceToken;
		}

		internal InterpolationSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 5;
			PunctuationSyntax punctuationSyntax = (PunctuationSyntax)reader.ReadValue();
			if (punctuationSyntax != null)
			{
				AdjustFlagsAndWidth(punctuationSyntax);
				_openBraceToken = punctuationSyntax;
			}
			ExpressionSyntax expressionSyntax = (ExpressionSyntax)reader.ReadValue();
			if (expressionSyntax != null)
			{
				AdjustFlagsAndWidth(expressionSyntax);
				_expression = expressionSyntax;
			}
			InterpolationAlignmentClauseSyntax interpolationAlignmentClauseSyntax = (InterpolationAlignmentClauseSyntax)reader.ReadValue();
			if (interpolationAlignmentClauseSyntax != null)
			{
				AdjustFlagsAndWidth(interpolationAlignmentClauseSyntax);
				_alignmentClause = interpolationAlignmentClauseSyntax;
			}
			InterpolationFormatClauseSyntax interpolationFormatClauseSyntax = (InterpolationFormatClauseSyntax)reader.ReadValue();
			if (interpolationFormatClauseSyntax != null)
			{
				AdjustFlagsAndWidth(interpolationFormatClauseSyntax);
				_formatClause = interpolationFormatClauseSyntax;
			}
			PunctuationSyntax punctuationSyntax2 = (PunctuationSyntax)reader.ReadValue();
			if (punctuationSyntax2 != null)
			{
				AdjustFlagsAndWidth(punctuationSyntax2);
				_closeBraceToken = punctuationSyntax2;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_openBraceToken);
			writer.WriteValue(_expression);
			writer.WriteValue(_alignmentClause);
			writer.WriteValue(_formatClause);
			writer.WriteValue(_closeBraceToken);
		}

		static InterpolationSyntax()
		{
			CreateInstance = (ObjectReader o) => new InterpolationSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(InterpolationSyntax), (ObjectReader r) => new InterpolationSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.InterpolationSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _openBraceToken, 
				1 => _expression, 
				2 => _alignmentClause, 
				3 => _formatClause, 
				4 => _closeBraceToken, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new InterpolationSyntax(base.Kind, newErrors, GetAnnotations(), _openBraceToken, _expression, _alignmentClause, _formatClause, _closeBraceToken);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new InterpolationSyntax(base.Kind, GetDiagnostics(), annotations, _openBraceToken, _expression, _alignmentClause, _formatClause, _closeBraceToken);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitInterpolation(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/InterpolationSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class InterpolationSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolatedStringContentSyntax
		Friend ReadOnly _openBraceToken As PunctuationSyntax

		Friend ReadOnly _expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax

		Friend ReadOnly _alignmentClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax

		Friend ReadOnly _formatClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax

		Friend ReadOnly _closeBraceToken As PunctuationSyntax

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property AlignmentClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax
			Get
				Return Me._alignmentClause
			End Get
		End Property

		Friend ReadOnly Property CloseBraceToken As PunctuationSyntax
			Get
				Return Me._closeBraceToken
			End Get
		End Property

		Friend ReadOnly Property Expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax
			Get
				Return Me._expression
			End Get
		End Property

		Friend ReadOnly Property FormatClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax
			Get
				Return Me._formatClause
			End Get
		End Property

		Friend ReadOnly Property OpenBraceToken As PunctuationSyntax
			Get
				Return Me._openBraceToken
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal openBraceToken As PunctuationSyntax, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax, ByVal alignmentClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax, ByVal formatClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax, ByVal closeBraceToken As PunctuationSyntax)
			MyBase.New(kind)
			MyBase._slotCount = 5
			MyBase.AdjustFlagsAndWidth(openBraceToken)
			Me._openBraceToken = openBraceToken
			MyBase.AdjustFlagsAndWidth(expression)
			Me._expression = expression
			If (alignmentClause IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(alignmentClause)
				Me._alignmentClause = alignmentClause
			End If
			If (formatClause IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(formatClause)
				Me._formatClause = formatClause
			End If
			MyBase.AdjustFlagsAndWidth(closeBraceToken)
			Me._closeBraceToken = closeBraceToken
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal openBraceToken As PunctuationSyntax, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax, ByVal alignmentClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax, ByVal formatClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax, ByVal closeBraceToken As PunctuationSyntax, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind)
			MyBase._slotCount = 5
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(openBraceToken)
			Me._openBraceToken = openBraceToken
			MyBase.AdjustFlagsAndWidth(expression)
			Me._expression = expression
			If (alignmentClause IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(alignmentClause)
				Me._alignmentClause = alignmentClause
			End If
			If (formatClause IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(formatClause)
				Me._formatClause = formatClause
			End If
			MyBase.AdjustFlagsAndWidth(closeBraceToken)
			Me._closeBraceToken = closeBraceToken
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal openBraceToken As PunctuationSyntax, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax, ByVal alignmentClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax, ByVal formatClause As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax, ByVal closeBraceToken As PunctuationSyntax)
			MyBase.New(kind, errors, annotations)
			MyBase._slotCount = 5
			MyBase.AdjustFlagsAndWidth(openBraceToken)
			Me._openBraceToken = openBraceToken
			MyBase.AdjustFlagsAndWidth(expression)
			Me._expression = expression
			If (alignmentClause IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(alignmentClause)
				Me._alignmentClause = alignmentClause
			End If
			If (formatClause IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(formatClause)
				Me._formatClause = formatClause
			End If
			MyBase.AdjustFlagsAndWidth(closeBraceToken)
			Me._closeBraceToken = closeBraceToken
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 5
			Dim punctuationSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			If (punctuationSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(punctuationSyntax)
				Me._openBraceToken = punctuationSyntax
			End If
			Dim expressionSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)
			If (expressionSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(expressionSyntax)
				Me._expression = expressionSyntax
			End If
			Dim interpolationAlignmentClauseSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationAlignmentClauseSyntax)
			If (interpolationAlignmentClauseSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(interpolationAlignmentClauseSyntax)
				Me._alignmentClause = interpolationAlignmentClauseSyntax
			End If
			Dim interpolationFormatClauseSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationFormatClauseSyntax)
			If (interpolationFormatClauseSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(interpolationFormatClauseSyntax)
				Me._formatClause = interpolationFormatClauseSyntax
			End If
			Dim punctuationSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.PunctuationSyntax)
			If (punctuationSyntax1 IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(punctuationSyntax1)
				Me._closeBraceToken = punctuationSyntax1
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitInterpolation(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InterpolationSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Select Case i
				Case 0
					greenNode = Me._openBraceToken
					Exit Select
				Case 1
					greenNode = Me._expression
					Exit Select
				Case 2
					greenNode = Me._alignmentClause
					Exit Select
				Case 3
					greenNode = Me._formatClause
					Exit Select
				Case 4
					greenNode = Me._closeBraceToken
					Exit Select
				Case Else
					greenNode = Nothing
					Exit Select
			End Select
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._openBraceToken, Me._expression, Me._alignmentClause, Me._formatClause, Me._closeBraceToken)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.InterpolationSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._openBraceToken, Me._expression, Me._alignmentClause, Me._formatClause, Me._closeBraceToken)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._openBraceToken, IObjectWritable))
			writer.WriteValue(DirectCast(Me._expression, IObjectWritable))
			writer.WriteValue(DirectCast(Me._alignmentClause, IObjectWritable))
			writer.WriteValue(DirectCast(Me._formatClause, IObjectWritable))
			writer.WriteValue(DirectCast(Me._closeBraceToken, IObjectWritable))
		End Sub
	End Class
End Namespace